<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Experimental design</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link href="libs/remark-css/rladies-fonts.css" rel="stylesheet" />
    <link href="libs/remark-css/hygge.css" rel="stylesheet" />
    <link rel="stylesheet" href="css/FMstyles.css" type="text/css" />
    <link rel="stylesheet" href="css/animate.css" type="text/css" />
    <link rel="stylesheet" href="css/my-theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, title-slide

.title[
# Experimental design
]
.author[
### Charlotte Soneson (<a href="mailto:charlotte.soneson@fmi.ch" class="email">charlotte.soneson@fmi.ch</a>)</br>
<p align="center">
<a href="https://imbeimainz.github.io/GTIPI2022"><img src="images/gtipi_logo.png" alt="" height="150"/></a>
</p>
]
.date[
### <p>2022/06/01</br> <a href="https://twitter.com/CSoneson"><code>&lt;svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"&gt;  &lt;path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"&gt;&lt;/path&gt;&lt;/svg&gt; @CSoneson</code></a></p>
]

---


layout: true



---

# Overview

.Large[
* Conceptual overview of experiments
* Vocabulary for discussing experimental design
* Factors influencing the measurements in an experiment, and strategies for reducing their influence
* More details on how to adjust for batch effects in practice
* Discussion about confounding
* Overview of different types of replicates
* Practical exercises to illustrate how experimental design choices affect the ability to draw conclusions from an experiment
]

---

# What is experimental design?

&lt;center&gt;
.full-width[.small-content-box-purple[
.Large["The organization of an experiment, to ensure that the **right type** of data,&lt;br&gt;and **enough** of it, is available&lt;br&gt;to answer the **questions of interest** as clearly and efficiently as possible."] 
]]
&lt;/center&gt;

--

&lt;center&gt;
.full-width[.small-content-box-purple[
.Large["Good experimental design begins with the end in mind."]
]]
&lt;/center&gt;

--

* .Large[What is my .red[hypothesis]?] 
* .Large[What am I going to .red[measure] (and how is this relevant to the hypothesis)?]
* .Large[What .red[other factors] affect this outcome measure (and am I controlling these appropriately)?]
* .Large[What groups do I want to .red[compare]?]

&lt;div class="my-footer"&gt;&lt;span&gt;http://www.stats.gla.ac.uk/steps/glossary/anova.html#expdes; Watkins (2016), doi:10.12746/swrccc2017.0517.226&lt;/span&gt;&lt;/div&gt;

???
* What is the population of interest?
* Importantly, the optimal design depends on the question that you would like to address!
  * Example: take care to include the right type of controls. When comparing groups, any differences between them will be picked up, whether they are caused by biology, technical/protocol differences or batch effects.
* Experimental design is intimately related to what you actually want to get out of your experiment.

---
background-image: url("images/be-more-explicit.png")
background-position: center
background-size: 90%
class: inverse

# A couple of motivating examples

---

# Case study: Lin et al (PNAS, 2014)

&lt;div class="pull-left" style="width: 40%"&gt;
&lt;center&gt;
.full-width[.content-box-yellow[
.large["In this study of a broad number of tissues between humans and mice, high-throughput sequencing assays on the transcriptome and epigenome reveal that, in general, differences dominate similarities between the two species."]&lt;br&gt;&lt;br&gt;
".large[[...] well beyond what was described previously, likely reflecting the fundamental physiological differences between these two organisms."]
]]
&lt;/center&gt;
&lt;/div&gt;

--

&lt;div class="pull-right" style="width: 58%"&gt;
&lt;center&gt;
![:scale 130%](images/Lin_Fig1c.png)

Subset of 13 mouse tissues and 13 human tissues with data generated at the same site (Stanford)
{{content}}
&lt;/center&gt;
&lt;/div&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Lin et al (2014), doi:10.1073/pnas.1413624111&lt;/span&gt;&lt;/div&gt;

???
ENCODE consortium, set out to address a difficult question
Results were a bit surprising, but argued that previous studies had been biased by using a small set of tissues, and focusing on tissue-specific genes etc.
"Overall, our results indicate that there is considerable RNA expression diversity between humans and mice, well beyond what was described previously, likely reflecting the fundamental physiological differences between these two organisms."

---

# Gilad &amp; Mizrahi-Man (F1000Research 2015)

&lt;div class="pull-left" style="width: 40%"&gt;
&lt;center&gt;
![:scale 100%](images/Mizrahi_Man_Fig2b.png)
&lt;/center&gt;
&lt;/div&gt;

--

&lt;div class="pull-right" style="width: 58%"&gt;
&lt;center&gt;
![:scale 100%](images/9f5f4330-d81d-46b8-9a3f-d8cb7aaf577e_figure1.gif)
&lt;/center&gt;
Species is .red[confounded] with sequencing machine.
{{content}}

--

&lt;ul&gt;
&lt;li&gt; Differences on the sample level 
  &lt;ul&gt;&lt;li&gt; human - heterogeneous set of deceased individuals (18-66 years), tissue flash frozen, RNA for three tissues acquired commercially.&lt;/li&gt;
  &lt;li&gt; mouse - single strain, similar age (10 weeks), unclear whether tissue was flash frozen.&lt;/li&gt;&lt;/ul&gt;&lt;/li&gt;
&lt;li&gt;Tissues not sex-matched&lt;/li&gt;
&lt;/ul&gt;
{{content}}

&lt;/div&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Gilad &amp; Mizrahi-Man (2015), doi:10.12688/f1000research.6536.1&lt;/span&gt;&lt;/div&gt;

???
Went into fastq files to extract info about sequencer, flow cell, lane
None of the possible confounders were reported in the paper
According to the authors, "22 of the libraries were constructed in a single batch and sequenced on four lanes"

---

# Follow-up experiments triggered

* Resequencing of a subset of the tissue samples with a non-confounded sequencing design - similar conclusions.
  * Doesn't address the underlying sample differences

&lt;center&gt;
![:scale 30%](images/resequencing_Lin_pca_1.png)![:scale 15%](images/resequencing_Lin_pca_2.png)
&lt;/center&gt;

--
* Analysis of Fantom5 CAGE data (matched human and mouse) - same conclusions. 
--

* Analysis of GTEx tissue samples
  * "Neither sex nor age contributed to the clustering in the first two principal components. Thus, these effects are likely to be small and have not been included in the studies of others as well."
  

???
The second set did not include ovary (exhausted material), different model sequencing instruments were used, and the reagents from the supplier were different. Also, the RNA may have changed over the two years since their extraction from tissue, especially for spleen and pancreas.

---

# Can we fix it in the analysis?

.pull-left[
![:scale 100%](images/Mizrahi_Man_Fig3b.png)
]

.pull-right[
- .large[Batch effect adjustment effectively "equalizes" or "aligns" the different batches.]
- .large[By construction, this removes systematic shift between different sequencing runs.]
- .large[**Unavoidably**, this removes also of the species effect (the species are sequenced in different batches).]
- .large[There's _no way_ to determine the source of the signal based on this data alone.]

]

???
Global normalization is not enough - batch effects are gene-specific. 
This is an annoying place to be - we've spent all the money collecting the data, and we can't say anything. Maybe the conclusions were right, maybe not. We'll need to collect more data to figure out what's happening and estimate the size of any confounding effect separately, and try to impose that on the original experiment.

---

# So - what can we conclude?

&lt;center&gt;
![:scale 80%](https://media.giphy.com/media/K6VhXtbgCXqQU/giphy.gif)
&lt;/center&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;https://media.giphy.com/media/K6VhXtbgCXqQU/giphy.gif&lt;/span&gt;&lt;/div&gt;

---

# Case study II (Larsen et al, BioMed Res Int 2014)

* Gene expression data from samples from five breast cancer subtypes, processed across six batches.
* Batch signal dominates in uncorrected data.

.pull-left[
&lt;center&gt;
![:scale 100%](images/Larsen-BioMedResInt-2014-Fig2a.png)
&lt;/center&gt;
]

.pull-right[
&lt;center&gt;
![:scale 100%](images/Larsen-BioMedResInt-2014-Fig2c.png)
&lt;/center&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Larsen et al (2014), doi:10.1155/2014/651751&lt;/span&gt;&lt;/div&gt;

???
This is a common way of diagnosing batch effects etc. Plot data, overlay any known info about the samples. 

---

# Case study II (Larsen et al, BioMed Res Int 2014)

.pull-left[
&lt;center&gt;
![:scale 100%](images/Larsen-BioMedResInt-2014-Fig2b.png)
&lt;/center&gt;
]

--

.pull-right[
&lt;center&gt;
![:scale 100%](images/Larsen-BioMedResInt-2014-Fig2d.png)
&lt;/center&gt;
]
{{content}}

--

* After batch correction, the subtypes represent the strongest signal in the data.
* This was **only** possible since the subtypes were spread randomly across the analysis batches!
* This illustrates that it's not (mainly) the presence of batch effects _per se_, but the confounding with the signal of interest, that causes problems for the analysis.
{{content}}

&lt;div class="my-footer"&gt;&lt;span&gt;Larsen et al (2014), doi:10.1155/2014/651751&lt;/span&gt;&lt;/div&gt;

---

# How do we 'adjust for' batch effects?

&lt;center&gt;
![:scale 100%](images/Larsen-BioMedResInt-2014-Fig2cd.png)
&lt;/center&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Larsen et al (2014), doi:10.1155/2014/651751&lt;/span&gt;&lt;/div&gt;

???
This is a common way of diagnosing batch effects etc. Plot data, overlay any known info about the samples. 

---

# How do we 'adjust for' batch effects in statistical modeling?
    
* .large[In statistical modeling (e.g., differential expression analysis), batch effects can be included as covariates (additional predictors) in the model.]

`$$y = \beta_0 + \beta_1\cdot batch + \beta_2\cdot condition + \varepsilon$$`

* .large[Note that if ] `\(batch\)` .large[ and ] `\(condition\)` .large[ coincide, we can at best hope to estimate ] `\(\beta_1 + \beta_2\)`.
--

* .large[For exploratory analysis, we often attempt to "eliminate" or "adjust for" such unwanted variation in advance, by subtracting the estimated effect from each variable (e.g. the expression of a gene).]
  * .large[Not recommended before statistical modelling, since the variance will be underestimated.]

--
* .large[Even partial confounding between batch and signal of interest can lead to problems.]

---



# Adjusting for batch effects - nonconfounded setup

&lt;center&gt;
![:scale 95%](images/batch_correction_noconfounding.png)
&lt;/center&gt;

---

# Adjusting for batch effects - confounded setup

&lt;center&gt;
![:scale 95%](images/batch_correction_confounding.png)
&lt;/center&gt;

---

# Adjusting for batch effects - partly confounded setup

&lt;center&gt;
![:scale 95%](images/batch_correction_partconfounding.png)
&lt;/center&gt;

---

# What can we learn from this?

- .large[Acknowledge that lots of factors can affect measurements in (biological) experiments]

&lt;center&gt;
![:scale 75%](images/Lazic_expeffects.png)
&lt;/center&gt;

--
- .large[We (typically) can't change the fact that these factors _do_ affect the measurements, so we have to adapt accordingly.]
--

- .large[Spending some extra time thinking through the experiment _before_ starting to collect the data can save a lot of time and headache later on.]
--

- .large[Always record as much information as possible about the samples, and share it with whoever analyzes the data. If you are the data analyst - ask questions!]
--

- .large[Avoid confounding between the factor of interest and any source of unwanted variation.]

&lt;div class="my-footer"&gt;&lt;span&gt;Lazic (2016), isbn:978-1107424883&lt;/span&gt;&lt;/div&gt;

???
We can't change the fact that these factors impact the measurements, we just need to design the experiment so that we can still say something about what we want
 (we'll talk about randomization and blocking later)

---

# Always beware of confounding

.Large[
* Careful design is always important - beware of confounding e.g. in situations where we
  * have to split a large data set into multiple analysis batches
  * use a control group from another (public) data set
  * use different reagent batches
  * collect an expanding data set (adding one condition at a time)
  * work in collaborative projects, where each center generates a subset of the data
  * change some aspect of the sample/library preparation protocol
  * work on an instrument with a drift/degradation over time
  * work with samples that degrade over time
  * ...
]

---

# So why should we care about experimental design? 

.pull-left[
.Large[
* Improved reproducibility of results
* Improved validity and generalizability of conclusions
* Efficient and ethical use of resources
  * Make sure that the design is adequate for answering the question of interest
  * Avoid having to repeat poorly designed experiments
  * Avoid spending resources on non-interpretable experiments
]
]

.pull-right[
![:scale 70%](images/Nature-reproducibility-graphic-online4-modified.jpg)

&lt;div class="my-footer"&gt;&lt;span&gt;Adapted from Baker (2016), doi:10.1038/533452a&lt;/span&gt;&lt;/div&gt; 

]

???
We'll talk about some strategies to avoid ending up in a situation where you are not able to address the question of interest with your experimental data.

---
background-image: url("images/field-experiment.jpeg")
background-position: center
background-size: cover
class: inverse

# Experimental design principles

---

# Different types of experiments

.pull-left[
.full-width[.content-box-red[
* **Learning** experiments
  * Discover as much as possible about a condition/situation/phenomenon, generate new hypotheses.
  * Example: "Does stress affect rodent behaviour?"
  * Heterogenous subjects and environment.
  * Several treatments/time points.
]]
]

--

.pull-right[
.full-width[.content-box-green[
* **Confirming** experiments
  * Verify/validate an observation (typically from a learning experiment or literature).
  * Example: "Does fox urine odour affect the amount of food Wistar rats consume during the first 24 hours after exposure?"
  * Standardized subjects and environment.
  * Small number of treatments/time points.
]]
]
{{content}}

--

* Both types of experiments are valuable.
* Most high-throughput experiments today are learning experiments.
* Beware of learning experiments that are _presented_ as confirming experiments.
  * Example: many protocols are tested, only the results from the "best" one are shown.
* Data from a learning experiment should not be reused in a follow-up confirming experiment.
{{content}}

&lt;div class="my-footer"&gt;&lt;span&gt;Adapted from Lazic (2016), isbn:978-1107424883; Sheiner (1997), doi:10.1016/S0009-9236(97)90160-0&lt;/span&gt;&lt;/div&gt; 

???
Look a bit more into the last point in the next session


---

# How to deal with the impact of unwanted variation?

&lt;br&gt;&lt;br&gt;&lt;br&gt;&lt;br&gt;
&lt;center&gt;
![:scale 100%](images/Lazic_expeffects.png)
&lt;/center&gt;

???
* Most of the time, our goal is to estimate treatment effects as well as possible. 
* (Technical) factors that are not of direct interest are often collectively referred to as 'batch effects'.
* We (typically) can't do much about the fact that all these factors _have_ an effect on our measurements.
* What we can (and should) do is to design our experiments in such a way that the impact on our ability to draw conclusions is minimized.

&lt;div class="my-footer"&gt;&lt;span&gt;Lazic (2016), isbn:978-1107424883&lt;/span&gt;&lt;/div&gt; 

---

# How to deal with the impact of unwanted variation?

.full-width[.small-content-box-gray[
* .large[**Minimize** it by holding factors constant (e.g., only considering female mice). This reduces unwanted variability, **but** we can't generalize our conclusions to groups that are not included!]
]]
--
.full-width[.small-content-box-gray[
* .large[Use **blocking** to balance the factor of interest across confounding factors. E.g., assign treatment group randomly within each sex group.] 
  * Extreme example: blocking _within each biological sample_ (e.g., take measurement from each mouse before and after a treatment). This leads to a **paired design**. 
]]
--
.full-width[.small-content-box-gray[
* .large[Use **randomization** to hopefully balance out the impact of any unknown confounder. Useful for unknown effects as well as effects that can only be measured at the end of the experiment.]
]]

&lt;div class="my-footer"&gt;&lt;span&gt;Adapted from https://stanlazic.github.io/pres/2018_Lazic_reproducibility.pdf&lt;/span&gt;&lt;/div&gt; 

---

# Example

.large[
* Let's say that you are collecting data from four untreated and four treated mice. 
* You can only collect and prepare data from four mice per day/session.
* How do you split your mice? Why? 
  * All replicates of one treatment in the same session to maximize the similarity among replicates
  * Completely random assignment
  * Replicates with each treatment evenly distributed between the two sessions
]

![](images/confounding_options.001.jpeg)

---

background-image: url("images/minions.jpg")
background-position: center
background-size: cover
class: inverse

# Replication

---

# Why do we need replicates?

* .Large[Required to estimate within-group variability, which is then compared to the observed between-group difference in statistical tests.]

![](images/compare_variance.001.png)
---

# Not all replicates are equal!

.Large['Technical' vs 'biological' replicates]

&lt;center&gt;
![:scale 70%](images/replicates.png)
&lt;/center&gt;

--

.Large[Common terminology, but sometimes confusing/not specific enough.]

&lt;div class="my-footer"&gt;&lt;span&gt;Klaus (2015), doi:10.15252/embj.201592958&lt;/span&gt;&lt;/div&gt; 

---

# Another classification approach

.pull-left[
.full-width[.content-box-green[
* .Large[.red[Biological units] (BU) - entities we want to make inferences about (e.g., animal, person).]
* .Large[.red[Experimental units] (EU) - the smallest entities that can be independently assigned to a treatment (e.g., animal, litter, cage, well).]
* .Large[.red[Observational units] (OU) - entities at which measurements are made.]
]]
]

.pull-right[
![:scale 100%](images/Lazic_expunits.png)
]

&lt;div class="my-footer"&gt;&lt;span&gt;Lazic (2016), isbn:978-1107424883&lt;/span&gt;&lt;/div&gt; 

---

# Another classification approach

.pull-left[
.full-width[.content-box-green[
* .Large[.red[Biological units] (BU) - entities we want to make inferences about (e.g., animal, person).]
* .Large[.red[Experimental units] (EU) - the smallest entities that can be independently assigned to a treatment (e.g., animal, litter, cage, well).]
* .Large[.red[Observational units] (OU) - entities at which measurements are made.]
]]
]
.pull-right[
* .Large[Replication of .red[biological units] is required to make a general statement of the effect of a treatment - you can't draw a conclusion about the effect of a compound on people in general by observing the effect in one person.] 
* .Large[Only replication of .red[experimental units] is true replication (independent error terms).]
]

---

# Not all replicates are equal

![](images/slope_hypothesis_testing_2x.png)

&lt;div class="my-footer"&gt;&lt;span&gt;https://xkcd.com/2533/&lt;/span&gt;&lt;/div&gt; 

---

# Experimental design checklist

&lt;input type="checkbox" unchecked&gt; I have a well-defined scientific question or hypothesis to investigate (recall learning vs confirming experiments).&lt;/input&gt;
&lt;br&gt;
&lt;input type="checkbox" unchecked&gt; I have thought about how the data will be analyzed (e.g., which comparisons will be done, and what can be learnt from each of them). In particular, I have considered what a suitable 'control' group is for the effect I am interested in.&lt;/input&gt;
&lt;br&gt;
&lt;input type="checkbox" unchecked&gt; The groups that I am planning to compare differ only in the aspect I am interested in investigating. Other confounders (technical/protocol differences, batch effects) are accounted for to the fullest extent possible.&lt;/input&gt;
&lt;br&gt;
&lt;input type="checkbox" unchecked&gt;I have considered which biological or technical factors that might influence my measurements, and decided how to handle them (ignoring, eliminating, blocking, randomizing). Sources of (unwanted) variation are not confounded with the effect of interest.&lt;/input&gt;
&lt;br&gt;
&lt;input type="checkbox" unchecked&gt; I have a suitable type and amount of replication, which allows me to generalize the results to the desired population.&lt;/input&gt;
&lt;br&gt;&lt;br&gt;&lt;br&gt;
&lt;center&gt;
.full-width[.content-box-purple[
.Large[Always record and report as much information as possible.]
]]
&lt;/center&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="css/my-macros.js"></script>
<script src="https://platform.twitter.com/widgets.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
